File: scripts/validate-verification.py
#!/usr/bin/env python3
"""Validate a moonbit-verify JSON artifact against its repository contract."""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SCHEMA_PATH = ROOT / "schemas" / "moonbit-verification.schema.json"
REQUIRED = {"schema_version", "status", "project_type", "targets", "checks", "started_at", "finished_at"}
STATUSES = {"pass", "blocked", "skipped"}
PROJECT_TYPES = {"lib", "cli", "ffi", "wasm", "parser", "async"}
CHECK_KEY = re.compile(r"^[A-Z][0-9]+$")


def load_json(path: Path) -> tuple[object | None, str | None]:
    try:
        return json.loads(path.read_text(encoding="utf-8")), None
    except FileNotFoundError:
        return None, f"File not found: {path}"
    except json.JSONDecodeError as exc:
        return None, f"Invalid JSON at {path}:{exc.lineno}:{exc.colno}: {exc.msg}"


def validate_artifact(data: object, schema: dict) -> list[str]:
    if not isinstance(data, dict):
        return ["Root must be a JSON object"]

    errors: list[str] = []
    required = set(schema.get("required", REQUIRED))
    errors.extend(f"Missing required field: {name}" for name in sorted(required - set(data)))

    allowed = set(schema.get("properties", {}))
    errors.extend(f"Unknown field: {name}" for name in sorted(set(data) - allowed))
    if errors:
        return errors

    if not isinstance(data["schema_version"], int) or data["schema_version"] < 1:
        errors.append("schema_version must be an integer >= 1")
    if data["status"] not in STATUSES:
        errors.append(f"status must be one of {sorted(STATUSES)}")
    if data["project_type"] not in PROJECT_TYPES:
        errors.append(f"project_type must be one of {sorted(PROJECT_TYPES)}")
    if not isinstance(data["targets"], list) or not data["targets"] or any(not isinstance(target, str) for target in data["targets"]):
        errors.append("targets must be a non-empty array of strings")
    if not isinstance(data["checks"], dict) or not data["checks"]:
        errors.append("checks must be a non-empty object")

    for key, result in data["checks"].items() if isinstance(data["checks"], dict) else []:
        if not CHECK_KEY.fullmatch(key):
            errors.append(f"Invalid check key: {key}")
        if not isinstance(result, dict):
            errors.append(f"checks.{key} must be an object")
            continue
        for field in ("status", "command", "exit_code"):
            if field not in result:
                errors.append(f"checks.{key} missing field: {field}")
        if not isinstance(result.get("command"), str) or not result["command"].strip():
            errors.append(f"checks.{key}.command must be non-empty")
        exit_code = result.get("exit_code")
        if exit_code is not None and (not isinstance(exit_code, int) or isinstance(exit_code, bool)):
            errors.append(f"checks.{key}.exit_code must be an integer or null")
        status = result.get("status")
        if status not in STATUSES:
            errors.append(f"checks.{key}.status must be one of {sorted(STATUSES)}")
        if status == "skipped" and not result.get("reason"):
            errors.append(f"checks.{key} skipped result requires reason")
        if status == "blocked" and (not result.get("reason") or not result.get("details")):
            errors.append(f"checks.{key} blocked result requires reason and details")

    for field in ("started_at", "finished_at"):
        if not isinstance(data[field], str) or "T" not in data[field]:
            errors.append(f"{field} must be an ISO 8601 date-time string")

    return errors


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate a moonbit-verify JSON artifact")
    parser.add_argument("--file", "-f", required=True, help="Verification artifact JSON path")
    args = parser.parse_args()

    schema, schema_error = load_json(SCHEMA_PATH)
    if schema_error or not isinstance(schema, dict):
        print(f"ERROR: {schema_error or 'verification schema must be an object'}", file=sys.stderr)
        return 1

    data, data_error = load_json(Path(args.file))
    if data_error:
        print(f"ERROR: {data_error}", file=sys.stderr)
        return 1

    errors = validate_artifact(data, schema)
    if errors:
        print("Validation FAILED:", file=sys.stderr)
        for error in errors:
            print(f"  - {error}", file=sys.stderr)
        return 1

    print(f"Validation PASS: {args.file}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/moonbit-skills/scripts/validate-verification.mjs
#!/usr/bin/env node
/* Validate a moonbit-verify JSON artifact against its repository contract. */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const SCHEMA_PATH = path.join(ROOT, "schemas", "moonbit-verification.schema.json");
const REQUIRED = ["schema_version", "status", "project_type", "targets", "checks", "started_at", "finished_at"];
const STATUSES = new Set(["pass", "blocked", "skipped"]);
const PROJECT_TYPES = new Set(["lib", "cli", "ffi", "wasm", "parser", "async"]);
const CHECK_KEY = /^[A-Z][0-9]+$/;

/** Compute {line, col} (1-based) for a character offset in text. */
function lineColAt(text, offset) {
  const before = text.slice(0, offset);
  const lines = before.split("\n");
  return { lineno: lines.length, colno: lines[lines.length - 1].length + 1 };
}

/**
 * Load a JSON file.
 * @returns {[data|null, error|null]}
 */
function loadJson(filePath) {
  try {
    return [JSON.parse(fs.readFileSync(filePath, "utf-8")), null];
  } catch (err) {
    if (err.code === "ENOENT") {
      return [null, `File not found: ${filePath}`];
    }
    if (err instanceof SyntaxError) {
      const m = /position (\d+)/.exec(err.message);
      if (m) {
        const pos = Number(m[1]);
        const { lineno, colno } = lineColAt(fs.readFileSync(filePath, "utf-8"), pos);
        return [null, `Invalid JSON at ${filePath}:${lineno}:${colno}: ${err.message}`];
      }
      return [null, `Invalid JSON at ${filePath}: ${err.message}`];
    }
    return [null, `Invalid JSON at ${filePath}: ${err.message}`];
  }
}

/**
 * Validate an artifact object against a schema.
 * @returns {string[]} list of errors (empty when valid)
 */
function validateArtifact(data, schema) {
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    return ["Root must be a JSON object"];
  }

  const errors = [];
  const required = new Set(schema?.required ?? REQUIRED);
  for (const name of [...required].sort()) {
    if (!(name in data)) errors.push(`Missing required field: ${name}`);
  }

  const allowed = new Set(Object.keys(schema?.properties ?? {}));
  for (const name of Object.keys(data).sort()) {
    if (!allowed.has(name)) errors.push(`Unknown field: ${name}`);
  }
  if (errors.length) return errors;

  if (!Number.isInteger(data.schema_version) || data.schema_version < 1) {
    errors.push("schema_version must be an integer >= 1");
  }
  if (!STATUSES.has(data.status)) {
    errors.push(`status must be one of ${[...STATUSES].sort().join(", ")}`);
  }
  if (!PROJECT_TYPES.has(data.project_type)) {
    errors.push(`project_type must be one of ${[...PROJECT_TYPES].sort().join(", ")}`);
  }
  if (
    !Array.isArray(data.targets) ||
    data.targets.length === 0 ||
    data.targets.some((t) => typeof t !== "string")
  ) {
    errors.push("targets must be a non-empty array of strings");
  }
  if (typeof data.checks !== "object" || data.checks === null || Array.isArray(data.checks) || Object.keys(data.checks).length === 0) {
    errors.push("checks must be a non-empty object");
  }

  if (typeof data.checks === "object" && data.checks !== null && !Array.isArray(data.checks)) {
    for (const key of Object.keys(data.checks)) {
      const result = data.checks[key];
      if (!CHECK_KEY.test(key)) {
        errors.push(`Invalid check key: ${key}`);
      }
      if (typeof result !== "object" || result === null || Array.isArray(result)) {
        errors.push(`checks.${key} must be an object`);
        continue;
      }
      for (const field of ["status", "command", "exit_code"]) {
        if (!(field in result)) errors.push(`checks.${key} missing field: ${field}`);
      }
      if (typeof result.command !== "string" || result.command.trim() === "") {
        errors.push(`checks.${key}.command must be non-empty`);
      }
      const exitCode = result.exit_code;
      if (exitCode !== null && exitCode !== undefined && (!Number.isInteger(exitCode) || typeof exitCode === "boolean")) {
        errors.push(`checks.${key}.exit_code must be an integer or null`);
      }
      const status = result.status;
      if (!STATUSES.has(status)) {
        errors.push(`checks.${key}.status must be one of ${[...STATUSES].sort().join(", ")}`);
      }
      if (status === "skipped" && !result.reason) {
        errors.push(`checks.${key} skipped result requires reason`);
      }
      if (status === "blocked" && (!result.reason || !result.details)) {
        errors.push(`checks.${key} blocked result requires reason and details`);
      }
    }
  }

  for (const field of ["started_at", "finished_at"]) {
    if (typeof data[field] !== "string" || !data[field].includes("T")) {
      errors.push(`${field} must be an ISO 8601 date-time string`);
    }
  }

  return errors;
}

function usage() {
  console.error("usage: validate-verification.mjs --file <path>");
}

function main() {
  const args = process.argv.slice(2);
  let fileArg = null;
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--file" || a === "-f") {
      fileArg = args[++i];
    } else if (a.startsWith("--file=")) {
      fileArg = a.slice("--file=".length);
    } else {
      usage();
      return 2;
    }
  }
  if (!fileArg) {
    usage();
    return 2;
  }

  const [schema, schemaError] = loadJson(SCHEMA_PATH);
  if (schemaError || typeof schema !== "object" || schema === null || Array.isArray(schema)) {
    console.error(`ERROR: ${schemaError || "verification schema must be an object"}`);
    return 1;
  }

  const [data, dataError] = loadJson(fileArg);
  if (dataError) {
    console.error(`ERROR: ${dataError}`);
    return 1;
  }

  const errors = validateArtifact(data, schema);
  if (errors.length) {
    console.error("Validation FAILED:");
    for (const error of errors) {
      console.error(`  - ${error}`);
    }
    return 1;
  }

  console.log(`Validation PASS: ${fileArg}`);
  return 0;
}

process.exitCode = main();
